printer function addon

Pikokosan opened the pull request:
added printer function. TODO: add printer selection to setup.py. currently defaults to printer #2 as printer #1 is VNC printer on test setup.

laurentalacoque commented on scripts/user_interface.py:
```diff
@@ -719,15 +734,30 @@ def onEnter(*args):
+        try:
+            conn = cups.Connection()
+            printers = conn.getPrinters()
+            default_printer = printers.keys()[1]#defaults to the first printer installed
```
I think this should be configurable in setup.py
Ideas to do so in `setup.py`:
- call conn.getPrinters() in setup and list available printers (best)
- or, at least, create a configuration key 'printer_name' filled by the user during setup and iterate on printers.keys to find it

What do you think ?

laurentalacoque commented on scripts/user_interface.py:
```diff
@@ -719,15 +734,30 @@ def onEnter(*args):
+            conn = cups.Connection()
+            printers = conn.getPrinters()
+            default_printer = printers.keys()[1]#defaults to the first printer installed
+            cups.setUser('pi')
```
Same as above, we should make the user at least a constant in `constants.py`

laurentalacoque commented on scripts/constants.py:
```diff
@@ -82,6 +82,7 @@
 COUNTDOWN_IMAGE_MAX_HEIGHT_RATIO = 0.2 #(0. - 1.) Ratio of the countdown images over screen height
 
 EMAIL_BUTTON_IMG  = os.path.join("ressources","ic_email.png")
+PRINT_BUTTON_IMG  = os.path.join("ressources","ic_print.png")
```
Did you include `ic_print.png` in this pull request?

Pikokosan replied:
I was using my own icon's at the time. I have a new icon to match the current still in the last commit.

laurentalacoque commented on scripts/setup.py:
```diff
@@ -49,47 +49,53 @@ def __init__(self,config,*args,**kwargs):
             self.want_email_cb  = Checkbutton(self.main_frame, text="Enable Email sending", variable=self.want_email_var, anchor=W, font='Helvetica')
             self.want_upload_cb  = Checkbutton(self.main_frame, text="Enable photo upload", variable=self.want_upload_var, anchor=W, font='Helvetica')
-            
+            self.want_print_cb = Checkbutton(self.main_frame, text="Enable photo print", variable=self.want_print_var, anchor=W, font='Helvetica')
```
Nice!
Next step would be to add a configuration page for the printer selection

laurentalacoque commented on scripts/setup.py:
```diff
@@ -13,6 +13,7 @@
 import configuration
 import constants
 import mykb
+import cups
```
Add a try/except/pass block to avoid breaking usage for people that don't need printing feature and haven't installed pycups

laurentalacoque commented on scripts/configuration.py:
```diff
@@ -19,23 +19,24 @@ class Configuration():
     full_screen  = True #Start application in full screen
     enable_email  = True #Enable the 'send email" feature
     enable_upload = True #Enable the upload feature
+    enable_print = True #Enable the printer feature
```
Should be False by default (failproof)
Also, a `selected_printer` default value (of `None`?) is needed here

Pikokosan commented:
Hey @laurentalacoque 
Here are update based on what you requested.

- Main concerns
  - check for pycups installations. if it's not installed it will remove the option for printing.
  - added note about printer drive setup being a option.
  - added more fail safes.


- Configuration
  - Added auto detection of current user for the printer. this uses getpass.
  - Setup.py automatically redraws page 0 to include a listbox for printer selection.
  - Changed printer option default to False.

I have also added everything to the setup.py for graphical setup and command line using the -c option. I have tested both the graphic and command line setup with pycups installed and uninstalled.

laurentalacoque commented on scripts/setup.py:
```diff
@@ -49,47 +57,89 @@ def __init__(self,config,*args,**kwargs):
+                        self.__erase_page()
+                        self.widgets.pop(0)
+                        self.widgets.insert(0,[self.want_email_cb, self.want_upload_cb, self.want_print_cb,self.use_soft_keyboard_cb,self.use_print_list])
+                        conn = cups.Connection()
```
This fails on my system because cups was not set up.
As soon as I check the printer feature, the setup screen goes blank and an error appears in the console (see screenshots below)

Simplest way to fix would be to try/catch this and in case of failure:
- Alert user that he/she must configure cups (possibly linking to a fix webpage)
- set the print option to false (and disabled)

![before_option_select](https://user-images.githubusercontent.com/16169826/43969976-2bb68e0a-9ccc-11e8-839c-5db68ef926d5.png)
![after_option_select](https://user-images.githubusercontent.com/16169826/43969980-30415e50-9ccc-11e8-84a4-718015ff14f3.png)

laurentalacoque commented on scripts/setup.py:
```diff
@@ -727,9 +778,28 @@ def console_assistant():
+        config.enable_print = ask_boolean("Do you want the 'Send photo to printer' feature?", config.enable_print)
+        print ""
+        if config.enable_print == True:
+            conn = cups.Connection()
```
Same here, this fails on my system and exits the setup program with error : 
```
Do you want the 'send photo by email' feature? [Y/n] => 

Do you want the 'auto-upload photos' feature? [Y/n] => 

Do you want the 'Send photo to printer' feature? [N/y] => y

Traceback (most recent call last):
  File "setup.py", line 1038, in <module>
    console_assistant()
  File "setup.py", line 785, in console_assistant
    conn = cups.Connection()
RuntimeError: failed to connect to server
(anew)laurent@localhost:~/pikosanTouchSelfie/TouchSelfie/scripts$
```

Pikokosan commented:
I'll look at it again tonight when I get home and add the try exception for the connection error.
